benchadapt/python/README.md

```diff
@@ -1,3 +1,132 @@
+
+1. How to get results, including what `command` should be (though it can be defined later
+by the user, if desired) and how to get the raw results (e.g. if they are in a file or
+directory of files, where they are and how to read them in).
+2. How to transform the results into a list of `BenchmarkResult` instances ((2) above) in
+the `._transform_results()` method.
+
+(3) and (4) are handled by the abstract class.
+
+Various adapters are alrady defined in the `adapters` submodule, including ones for
+Google Benchmark and Folly, as well as a generic `CallableAdapter`, which takes a Python
+`Callable` instance (a function or class with a `__call__()` method) that returns a list
+of `BenchmarkResult` instances directly instead of a shell command. Many more adapters
+are possible; if you create one corresponding to a benchmarking tool, please make a PR!
+
+#### Running an adapter
+
+Adapters have separate `.run()` and `.publish_results()` methods; the former runs the
+benchmarks, transforms the results, and stores them in a `.results` attribute of the
+instance. It does not post them, so is useful for looking at results interactively before
+sending them. `.publish_results()` takes the results from the `.results` attribute and
+posts them to a Conbench server.
+
+The whole instance also has a `__call__()` method defined so it can be called like a
+function that both runs and publishes, so a somewhat minimal script for running
+benchmarks in CI might look like
+
+``` python
+import os
+
+from benchadapt.adapters import GoogleBenchmarkAdapter
+
+adapter = GoogleBenchmarkAdapter(
+    command=["bash", "./run-benchmarks.sh"],
+    result_file="benchmarks.json",
+    result_fields_override={
+        "run_reason": os.getenv("CONBENCH_RUN_REASON")
+    },
+    result_fields_append={
+        "info": {"build_version": os.getenv("MY_BUILD_VERSION")},
+        "context": {"compiler_flags": os.getenv("MY_COMPILER_FLAGS")}
+    }
+)
+adapter()
+```
+
+Of note:
+
+- `result_fields_override` will replace the whole attribute with a new value. This works
+with all types (strings, dicts, etc.), so here `run_reason` will be set for all results.
+- `result_fields_append` will append the new values to dicts which may already have data.
+Here, `build_version` will be appended to the `info` dict. In this case it is an empty
+dict anyway, so this is equivalent to
+`result_fields_override={"info": {"build_version": os.getenv("MY_BUILD_VERSION")}})`.
+But the `context` dict will already contain a `"benchmark_language"` key; this will be
+retained, and `compiler_flags` will be appended.
+- For this to work, a lot of environment variables have to be set! This includes ones
+with information about the Conbench server and the current git metadata. See the
+"Environment Variables" section below for a full list.
+
+
+## Environment variables
+
+benchadapt relies on a number of environment variables. The Conbench server ones
```

Comment:
> benchadapt relies on a number of environment variables

Can we say "the `benchadapt CLI` or the `benchadapt library` here, depending on what's correct?

Reply:
>  The Conbench server ones

I think this is confusing. Does a consumer of these docs need to know that the Conbench webapp can be configured with environment variables that have the same names?

Reply:
There is no benchadapt CLI; for a certain use-case, that's benchconnect. benchadapt is just a library.